src/mechanics/tools/QuasistaticSolver.cpp

```diff
@@ -0,0 +1,117 @@
+#include "mechanics/tools/QuasistaticSolver.h"
+#include "math/EigenSparseSolve.h"
+#include "math/NewtonRaphson.h"
+
+using namespace NuTo;
+
+QuasistaticSolver::QuasistaticSolver(TimeDependentProblem& s, DofType dof)
+    : mProblem(s)
+    , mDof(dof)
+{
+}
+
+void QuasistaticSolver::SetConstraints(Constraint::Constraints constraints)
+{
+    mConstraints = constraints;
+    if (mX.rows() == 0)
+        mX = mProblem.RenumberDofs(constraints, {mDof}, GlobalDofVector()).J[mDof];
+    else
+        mX = mProblem.RenumberDofs(constraints, {mDof}, ToGlobalDofVector(mX)).J[mDof];
+
+    mCmat = constraints.BuildConstraintMatrix(mDof, mX.rows());
+}
+
+void QuasistaticSolver::SetGlobalTime(double globalTime)
+{
+    mTimeStep = globalTime - mGlobalTime;
+    mGlobalTime = globalTime;
+}
+
+std::pair<Eigen::SparseMatrix<double>, Eigen::VectorXd>
+QuasistaticSolver::TrialSystem(const Eigen::VectorXd& x, double globalTime, double timeStep)
```

Comment:
Why do you need global time and timeStep, can't you compute the timestep from the (new) global time and the time stored during the last update? I also would not write a setTime function, only change the time when an update is made.